cmd/gaiacli/README.md

```diff
@@ -0,0 +1,195 @@
+```
+
+### Connection Creation
+
+In order to send transactions using IBC there are two differnt handshakes that must be preformed. First there is a `connection` created between the two chains. Once the connection is created, an application specific `channel` handshake is preformed which allows the transfer of application specific data. Examples of applications are token transfer, cross-chain validation, cross-chain accounts, and in this tutorial `ibc-mock`.
+
+Create a `connection` with the following command:
+
+> NOTE: This command broadcasts a total of 7 transactions between the two chains from 2 different wallets. At the start of the command you will be prompted for passwords for the two different keys. The command may then take some time. Please wait for the command to return!
+
+```shell
+gaiacli \
+  --home ibc0/n0/gaiacli \
+  tx ibc connection handshake \
+  conn0 c0 $(gaiacli --home ibc1/n0/gaiacli q ibc client path) \
+  conn1 c1 $(gaiacli --home ibc0/n0/gaiacli q ibc client path) \
+  --chain-id2 ibc1 \
+  --from1 n0 --from2 n1 \
+  --node1 tcp://localhost:26657 \
+  --node2 tcp://localhost:26557
+```
+
+Once the connection is established you should be able to query it:
+
+```bash
+gaiacli --home ibc0/n0/gaiacli q ibc connection connection conn0 --indent --trust-node
+gaiacli --home ibc1/n0/gaiacli q ibc connection connection conn1 --indent --trust-node
+```
+
+### Channel
+
+Now that the `connection` has been created, its time to establish a `channel` for the `ibc-mock` application protocol. This will allow sending of data between `ibc0` and `ibc1`. To create the `channel`, run the following command:
+
+> NOTE: This command broadcasts a total of 7 transactions between the two chains from 2 different wallets. At the start of the command you will be prompted for passwords for the two different keys. The command may then take some time. Please wait for the command to return!
+
+```bash
+gaiacli \
+  --home ibc0/n0/gaiacli \
+  tx ibc channel handshake \
+  ibcmocksend chan0 conn0 \
+  ibcmockrecv chan1 conn1 \
+  --node1 tcp://localhost:26657 \
+  --node2 tcp://localhost:26557 \
+  --chain-id2 ibc1 \
+  --from1 n0 --from2 n1
+```
+
+You can query the `channel` after establishment by running the following command:
+
+```bash
+gaiacli --home ibc0/n0/gaiacli query ibc channel channel ibcmocksend chan0 --indent --trust-node
+gaiacli --home ibc1/n0/gaiacli query ibc channel channel ibcmockrecv chan1 --indent --trust-node
+```
+
+## Send Packet
+
+To send a packet using the `ibc-mock` application protocol, you need to know the channel you plan to send on, as well as the sequence number on the channel. To get the sequence you use the following commands:
+
+```bash
+# Returns the last sequence number
+gaiacli --home ibc0/n0/gaiacli q ibcmocksend sequence chan0
+
+# Returns the next expected sequence number, for use in scripting
+gaiacli --home ibc0/n0/gaiacli q ibcmocksend next chan0
```

Comment:
On a fresh install:

Running `gaiacli --home ibc0/n0/gaiacli q ibcmocksend sequence chan0` returns `0`

Running `gaiacli --home ibc0/n0/gaiacli q ibcmocksend next chan0` immediately afterward returns `2` (without having sent any packets)

I was able to reproduce the above after stopping the testnet gaiad processes, removing the entire ibc-testnets directory, and starting over. Will try on a new machine tomorrow (currently, on a Mac).

Reply:
Wanted to add that manually inputting the correct sequence number (should be `1` instead of `2`) to successfully send a packet the first time, then using the above to get subsequent "next sequence" numbers works.

Reply:
https://github.com/cosmos/cosmos-sdk/blob/e69029151525422924e384cd9175778fa8533c51/x/ibc/mock/send/client/cli/query.go#L70

In L74 of the function add 1 to res, so if value is nil, res should be set to 0 instead of 1.